confmap/confmap.go

```diff
@@ -306,6 +307,21 @@ func isStringyStructure(t reflect.Type) bool {
 	return false
 }
 
+// safeWrapDecodeHookFunc wraps a DecodeHookFuncValue to ensure that it is safe to call all
+// methods in fromVal.
+//
+// Use this only if the hook does not need to be called on nil values.
+func safeWrapDecodeHookFunc(
+	f mapstructure.DecodeHookFuncValue,
+) mapstructure.DecodeHookFuncValue {
+	return func(fromVal reflect.Value, toVal reflect.Value) (any, error) {
+		if !fromVal.IsValid() {
+			return nil, nil
```

Comment:
Shouldn't this return the original? It may not be valid but it may have a type, and returning bare nil erases the type.

Reply:
There is no original in this case: we only hit this case with an untyped nil. See golang/go/issues/51649 and this example I wrote: https://go.dev/play/p/DrlAK2EZG7x

If we have a typed nil, that will be passed down to the hook and things will work fine